### PR TITLE
chore(dbm): disable propagation for dbapi fetch methods

### PR DESCRIPTION
**Pull request by mabdinur**
## Description

We recently added support for context propagation to database queries that are instrumented using `ddtrace.contrib.dbapi`. However, this change breaks the instrumentation for fetch methods (`fetchone`, `fetchmany`, and `fetchall`). We can safely disable propagation for these methods.

## Motivation

This PR ensure only `callproc()`, `execute()` and `executemany()` operations are used in DBM propagation. 

The following [dbapi operations](https://peps.python.org/pep-0249/#id20) do not execute queries and should not attempt to link APM Spans to Database Operations:
- Cusor.fetchone()
- Cusor.fetchmany()
- Cusor.fetchall()

## Checklist
- [ ] Add additional sections for `feat` and `fix` pull requests.
- [ ] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [x] Title is accurate.
- [x] Description motivates each change.
- [x] No unnecessary changes were introduced in this PR.
- [x] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [x] All relevant GitHub issues are correctly linked.
- [x] Backports are identified and tagged with Mergifyio.

### CODE REVIEW

File: ddtrace/contrib/dbapi/__init__.py [hunk 1]
```diff
@@ -100,7 +101,11 @@ def _trace_method(self, method, name, resource, extra_tags, *args, **kwargs):
             if not isinstance(self, FetchTracedCursor):
                 s.set_tag(ANALYTICS_SAMPLE_RATE_KEY, self._self_config.get_analytics_sample_rate())
 
-            args, kwargs = _inject_dbm_comment_in_query_arg(s, args, kwargs)
+            if dbm_operation:
+                # If the traced operation executes a database query (ex: cursor.execute(..) or cursor.executemany(..))
+                # then attempt to add DBM tags to the query.
+                args, kwargs = _propagate_dbm_comment(s, args, kwargs)
```
Comment (mabdinur) [inline ddtrace/contrib/dbapi/__init__.py:107]:
```suggestion
                args, kwargs = _propagate_dbm_context(s, args, kwargs)
```

File: ddtrace/contrib/dbapi/__init__.py [hunk 2]
```diff
@@ -305,7 +313,7 @@ def _get_module_name(conn):
     return conn.__class__.__module__.split(".")[0]
 
 
-def _inject_dbm_comment_in_query_arg(dbspan, args, kwargs):
+def _propagate_dbm_comment(dbspan, args, kwargs):
```
Comment (mabdinur) [inline ddtrace/contrib/dbapi/__init__.py:316]:
```suggestion
def _propagate_dbm_context(dbspan, args, kwargs):
```

**Review by ZStriker19 — CHANGES_REQUESTED** [review]

File: tests/contrib/dbapi/test_dbapi.py [hunk 3]
```diff
@@ -33,21 +33,31 @@ def test_execute_wrapped_is_called_and_returned(self):
...
-        traced_cursor = TracedCursor(cursor, Pin("pin_name", tracer=self.tracer), {})
-        query = "SELECT * FROM db;"
-        traced_cursor.execute(query)
+        traced_cursor = TracedCursor(cursor, Pin(service="orders-db", tracer=self.tracer), {})
+
+        # The following operations should not generate DBM comments
+        traced_cursor.callproc("proc")
+        cursor.callproc.assert_called_once_with("proc")
         spans = self.tracer.pop()
         assert len(spans) == 1
```
[the remark this answers is not in the document]
Comment (mabdinur) [reply tests/contrib/dbapi/test_dbapi.py:50]: The test is already doing this. Previously the dbm comment would have been prepended to the string passed to callproc(). This test asserts that the dbm comment is not added to “proc”.  Below we test that the execute and execute many operations do add the dbm comment.

**Review by ZStriker19 — DISMISSED** [review]

**Comment by majorgreys** [comment]
Can we propagate the dbm comment to a procedure call if the dbms support stored procedures?

**Comment by majorgreys** [comment]
At least in the case of psycopg2, it would be tricky to add the sql comment since that is done in the C code:

https://cs.github.com/psycopg/psycopg2/blob/3182ea2303f4c7fb52c0f17018ddf98aa03ab4c8/psycopg/cursor_type.c#L1116

File: ddtrace/contrib/psycopg/patch.py [hunk 4]
```diff
@@ -61,17 +61,53 @@ def unpatch():
 class Psycopg2TracedCursor(dbapi.TracedCursor):
     """TracedCursor for psycopg2"""
 
-    def _trace_method(self, method, name, resource, extra_tags, *args, **kwargs):
```
Comment (mabdinur) [inline ddtrace/contrib/psycopg/patch.py:64]: I can move this refactor to it's own PR. The issue I am trying to solve here is in pyscopg the query isn't always a string. The query can be a [Composable object](https://www.psycopg.org/docs/sql.html#psycopg2.sql.Composable). In pycopg we must convert the query to a string before calling `TracedCursor._trace_method()`

**Review by ZStriker19 — APPROVED** [review]

File: ddtrace/contrib/psycopg/patch.py (hunk 4 above)
Comment (majorgreys) [reply ddtrace/contrib/psycopg/patch.py:64]: Yes, move this to a separate PR.

Comment (mabdinur) [reply ddtrace/contrib/psycopg/patch.py:64]: https://github.com/DataDog/dd-trace-py/pull/4449

**Comment by mabdinur** [comment]
Blocked by: https://github.com/DataDog/dd-trace-py/pull/4449

File: ddtrace/contrib/dbapi/__init__.py [hunk 5]
```diff
@@ -305,15 +309,16 @@ def _get_module_name(conn):
...
-def _inject_dbm_comment_in_query_arg(dbspan, args, kwargs):
-    # type: (...) -> Tuple[Tuple[Any, ...], Dict[str, Any]]
+def _propagate_dbm_context(dbspan, args):
+    # type: (...) -> Tuple[Tuple[Any, ...]]
     dbm_comment = _database_monitoring._get_dbm_comment(dbspan)
     if dbm_comment is None:
-        return args, kwargs
-
-    # if dbm comment is not None prepend it to the query
-    query = get_argument_value(args, kwargs, 0, "query")
```
Comment (majorgreys) [inline ddtrace/contrib/dbapi/__init__.py:315]: Why not avoid injection if the arguments are not as we expect?

```
if not query:
     return args, kwargs
```

**Review by majorgreys — APPROVED** [review]
no blockers from me at this point

**Review by ZStriker19 — APPROVED** [review]

**Review by ZStriker19 — APPROVED** [review]

File: tests/contrib/dbapi/test_dbapi.py [hunk 6]
```diff
@@ -500,6 +506,43 @@ def test_callproc_can_handle_arbitrary_args(self):
...
         self.reset()
 
+    @TracerTestCase.run_in_subprocess(
+        env_overrides=dict(
+            DD_TRACE_SQL_COMMENT_INJECTION_MODE="full",
```
Comment (mabdinur) [inline tests/contrib/dbapi/test_dbapi.py:511]:
```suggestion
            DD_DBM_PROPAGATION_MODE ="full",
```

File: ddtrace/contrib/psycopg/patch.py [hunk 7]
```diff
@@ -61,12 +61,16 @@ def unpatch():
...
+    def _trace_method(self, method, name, resource, extra_tags, dbm_operation, *args, **kwargs):
         # treat psycopg2.sql.Composable resource objects as strings
         if isinstance(resource, Composable):
             resource = resource.as_string(self.__wrapped__)
+            # DBM context propagation is only supported for query args with the following type: string
+            dbm_operation = False
```
Comment (majorgreys) [inline ddtrace/contrib/psycopg/patch.py:69]: Does a test fail if this is not disabled?

Comment (mabdinur) [reply ddtrace/contrib/psycopg/patch.py:69]: Yup a few tests will fail. Here's an example: https://github.com/DataDog/dd-trace-py/blob/1.x/tests/contrib/psycopg/test_psycopg.py#L316

**Review by ZStriker19 — APPROVED** [review]

File: ddtrace/contrib/psycopg/patch.py (hunk 7 above)
Comment (majorgreys) [reply ddtrace/contrib/psycopg/patch.py:69]: Ok, we will want to revisit this as part of a clean up of the psycopg integration.

**Review by majorgreys — APPROVED** [review]

**Pull request merged**